lib/ansible/modules/cloud/amazon/s3_bucket.py

```diff
@@ -155,6 +156,95 @@ def create_tags_container(tags):
     return tags_obj
 
 
+def compare_policy_statement_part(current, new):
+    ''' Takes a matching part of a current policy statement and the
+        new policy statement and compares them. Ensures lists are
+        compared to other lists and calls itself recursively for dicts
+        so nesting structures are as expected.
+        Returns True if the policy has changed.
+    '''
+    if current != new:
+        if isinstance(current, dict):
+            if set(current.keys()) != set(new.keys()):
+                return True
+            for each in set(current.keys()):
+                current_inner = current[each]
+                new_inner = new[each]
+                # recursively checking each value in the dicts
+                if compare_policy_statement_part(current_inner, new_inner):
+                    return True
+        if isinstance(current, list) and isinstance(new, string_types):
+            current = [to_text(each) for each in current]
+            new = [to_text(new)]
+        elif isinstance(new, list) and isinstance(current, string_types):
+            new = [to_text(each) for each in new]
+            current = [to_text(current)]
+        if not isinstance(current, dict) and current != new:
+            return True
+
+    return False
+
+
+def sort_policy_statement(current_policy, new_policy):
+    ''' Takes two policies and looks for matching statement parts (since it is a list of dicts
+        and cannot be easily ordered).
+        Returns a list of tuple pairs that can then be given to
+        compare_policy_statement_part()
+        Return sample:
+          [({
+             "Sid":"AddCannedAcl",
+             "Effect":"Allow",
+             "Principal": {
+               "AWS": "arn:aws:iam::XXXXXXXXXXXX:user/name"
+             },
+             "Action":["s3:PutObjectAcl"],
+             "Resource": "arn:aws:s3:::XXXXXXXX/*"
+           },
+            {
+             "Sid": "AddCannedAcl",
+             "Effect": "Allow",
+             "Principal": {
+                "AWS": ["arn:aws:iam::XXXXXXXXXXXX:user/name"]
+             },
+             "Action": "s3:PutObjectAcl",
+             "Resource": ["arn:aws:s3:::XXXXXXXX/*"]
+           })]
```

Comment:
Thanks for reviewing this. I'm not quite sure how to implement this part of the review.
The statement of a current policy passed to this function may look something like:
```
[{	u'Action': [u's3:PutObjectAcl'],
   	u'Sid': u'AddCannedAcl',
   	u'Resource': u'arn:aws:s3:::bucketname/*’,
   	u'Effect': u'Allow',
   	u'Principal': {u'AWS': u'arn:aws:iam::XXXXXXXXXXXXX:user/username’}
},
 {	u'Action': [u's3:PutObject', u's3:PutObjectAcl'],
  	u'Sid': u’PutObjectAndAcl’,
  	u'Resource': u'arn:aws:s3:::bucketname/*’,
  	u'Effect': u'Allow',
  	u'Principal': {u'AWS': u'arn:aws:iam::XXXXXXXXXXXX:user/username’}
}]
```
I'm currently just looking for an identical SID in this function for each dict in the statement of the new policy and current policy; if there isn't one, it means there is definitely going to be a change made to the current policy so basically we don't need to turn things into lists to compare everything. I'm happy to add some more logic here though if you think it will fix something.